Spammy log when one bookie of ensemble is down

When one bookie of the ensemble is down, the others bookies client  spam in the log ***org.apache.bookkeeper.proto.PerChannelBookieClient*** this exception: 

```
20-03-10-10-42-01       Could not connect to bookie: [id: 0x823ef32e, L:/10.162.1.12:42330]/es-dev1.sviluppo.dna:3181, current state CONNECTING :
20-03-10-10-42-01       io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(..) failed: Connection refused: es-dev1.sviluppo.dna/10.162.1.11:3181
io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(..) failed: Connection refused: es-dev1.sviluppo.dna/10.162.1.11:3181
        at io.netty.channel.unix.Socket.finishConnect(..)(Unknown Source)
Caused by: io.netty.channel.unix.Errors$NativeConnectException: syscall:getsockopt(..) failed: Connection refused
        ... 1 more 

```
In our case, in 8m it produced 1,5gb log.
For riproduced it, i simple shutdown a bookie and checked **org.apache.bookkeeper.proto.PerChannelBookieClient**

--- a/bookkeeper-server/pom.xml
+++ b/bookkeeper-server/pom.xml
@@ -234,7 +234,7 @@
         <version>${maven-javadoc-plugin.version}</version>
         <configuration>
           <!-- Avoid for missing javadoc comments to be marked as errors -->
-          <doclint>-Xdoclint:none</doclint>
+          <doclint>none</doclint>
           <subpackages>org.apache.bookkeeper.client:org.apache.bookkeeper.conf:org.apache.bookkeeper.feature</subpackages>
           <groups>
             <group>

--- a/bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java
+++ b/bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java
@@ -190,6 +190,10 @@ public class ClientConfiguration extends AbstractConfiguration<ClientConfigurati
     // Registration Client
     protected static final String REGISTRATION_CLIENT_CLASS = "registrationClientClass";
 
+    // Logs
+    protected static final String CLIENT_CONNECT_BOOKIE_UNAVAILABLE_LOG_THROTTLING =
+            "clientConnectBookieUnavailableLogThrottling";
+
     /**
      * Construct a default client-side configuration.
      */
@@ -1925,6 +1929,28 @@ public class ClientConfiguration extends AbstractConfiguration<ClientConfigurati
      */
     public boolean getStoreSystemtimeAsLedgerCreationTime() {
         return getBoolean(STORE_SYSTEMTIME_AS_LEDGER_CREATION_TIME, false);
+    }
+
+    /**
+     * Set the log frequency when a bookie is unavailable, in order to limit log filesize.
+     *
+     * @param throttleValue
+     * @param unit
+     * @return client configuration.
+     */
+    public ClientConfiguration setClientConnectBookieUnavailableLogThrottling(
+            int throttleValue, TimeUnit unit) {
+        setProperty(CLIENT_CONNECT_BOOKIE_UNAVAILABLE_LOG_THROTTLING, unit.toMillis(throttleValue));
+        return this;
+    }
+
+    /**
+     * Get the log frequency when a bookie is unavailable, in milliseconds.
+     *
+     * @return log frequency when a bookie is unavailable, in milliseconds.
+     */
+    public long getClientConnectBookieUnavailableLogThrottlingMs() {
+        return getLong(CLIENT_CONNECT_BOOKIE_UNAVAILABLE_LOG_THROTTLING, 5_000L);
     }
 
     @Override

--- a/bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java
+++ b/bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java
@@ -83,6 +83,7 @@ import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 import java.util.function.BiPredicate;
 
+import java.util.function.Consumer;
 import javax.net.ssl.SSLHandshakeException;
 import javax.net.ssl.SSLPeerUnverifiedException;
 import lombok.SneakyThrows;
@@ -345,6 +346,7 @@ public class PerChannelBookieClient extends ChannelInboundHandlerAdapter {
     private final ExtensionRegistry extRegistry;
     private final SecurityHandlerFactory shFactory;
     private volatile boolean isWritable = true;
+    private long lastBookieUnavailableLog = 0;
 
     public PerChannelBookieClient(OrderedExecutor executor, EventLoopGroup eventLoopGroup,
                                   BookieId addr, BookieAddressResolver bookieAddressResolver) throws SecurityException {
@@ -2483,12 +2485,12 @@ public class PerChannelBookieClient extends ChannelInboundHandlerAdapter {
                     Throwable cause = future.cause();
                     if (cause instanceof UnknownHostException || cause instanceof NativeIoException) {
                         // Don't log stack trace for common errors
-                        LOG.warn("Could not connect to bookie: {}/{}, current state {} : {}",
-                                future.channel(), bookieId, state, future.cause().getMessage());
+                        logBookieUnavailable((nil) -> LOG.warn("Could not connect to bookie: {}/{}, current state {} : {}",
+                                future.channel(), bookieId, state, future.cause().getMessage()));
                     } else {
                         // Regular exceptions, include stack trace
-                        LOG.error("Could not connect to bookie: {}/{}, current state {} : ",
-                                future.channel(), bookieId, state, future.cause());
+                        logBookieUnavailable((nil) -> LOG.error("Could not connect to bookie: {}/{}, current state {} : ",
+                                future.channel(), bookieId, state, future.cause()));
                     }
 
                     rc = BKException.Code.BookieHandleNotAvailableException;
@@ -2516,6 +2518,14 @@ public class PerChannelBookieClient extends ChannelInboundHandlerAdapter {
             }
 
             makeWritable();
+        }
+
+        private void logBookieUnavailable(Consumer<Void> log) {
+            final long now = System.currentTimeMillis();
+            if ((now - lastBookieUnavailableLog) > 30_000) {
+                log.accept(null);
+                lastBookieUnavailableLog = now;
+            }
         }
     }
 

--- a/bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java
+++ b/bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java
@@ -83,7 +83,6 @@ import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 import java.util.function.BiPredicate;
 
-import java.util.function.Consumer;
 import javax.net.ssl.SSLHandshakeException;
 import javax.net.ssl.SSLPeerUnverifiedException;
 import lombok.SneakyThrows;
@@ -346,7 +345,7 @@ public class PerChannelBookieClient extends ChannelInboundHandlerAdapter {
     private final ExtensionRegistry extRegistry;
     private final SecurityHandlerFactory shFactory;
     private volatile boolean isWritable = true;
-    private long lastBookieUnavailableLog = 0;
+    private long lastBookieUnavailableLogTimestamp = 0;
 
     public PerChannelBookieClient(OrderedExecutor executor, EventLoopGroup eventLoopGroup,
                                   BookieId addr, BookieAddressResolver bookieAddressResolver) throws SecurityException {
@@ -2485,11 +2484,11 @@ public class PerChannelBookieClient extends ChannelInboundHandlerAdapter {
                     Throwable cause = future.cause();
                     if (cause instanceof UnknownHostException || cause instanceof NativeIoException) {
                         // Don't log stack trace for common errors
-                        logBookieUnavailable((nil) -> LOG.warn("Could not connect to bookie: {}/{}, current state {} : {}",
+                        logBookieUnavailable(() -> LOG.warn("Could not connect to bookie: {}/{}, current state {} : {}",
                                 future.channel(), bookieId, state, future.cause().getMessage()));
                     } else {
                         // Regular exceptions, include stack trace
-                        logBookieUnavailable((nil) -> LOG.error("Could not connect to bookie: {}/{}, current state {} : ",
+                        logBookieUnavailable(() -> LOG.error("Could not connect to bookie: {}/{}, current state {} : ",
                                 future.channel(), bookieId, state, future.cause()));
                     }
 
@@ -2520,11 +2519,11 @@ public class PerChannelBookieClient extends ChannelInboundHandlerAdapter {
             makeWritable();
         }
 
-        private void logBookieUnavailable(Consumer<Void> log) {
+        private void logBookieUnavailable(Runnable logger) {
             final long now = System.currentTimeMillis();
-            if ((now - lastBookieUnavailableLog) > 30_000) {
-                log.accept(null);
-                lastBookieUnavailableLog = now;
+            if ((now - lastBookieUnavailableLogTimestamp) > conf.getClientConnectBookieUnavailableLogThrottlingMs()) {
+                logger.run();
+                lastBookieUnavailableLogTimestamp = now;
             }
         }
     }

--- a/bookkeeper-stats/pom.xml
+++ b/bookkeeper-stats/pom.xml
@@ -34,7 +34,7 @@
         <version>${maven-javadoc-plugin.version}</version>
         <configuration>
           <!-- Avoid for missing javadoc comments to be marked as errors -->
-          <doclint>-Xdoclint:none</doclint>
+          <doclint>none</doclint>
           <subpackages>org.apache.bookkeeper.stats</subpackages>
           <groups>
             <group>

--- a/pom.xml
+++ b/pom.xml
@@ -870,11 +870,9 @@
         <artifactId>maven-javadoc-plugin</artifactId>
         <version>${maven-javadoc-plugin.version}</version>
         <configuration>
+          <additionalparam>-notimestamp</additionalparam>
           <!-- Prevent missing javadoc comments from being marked as errors -->
-          <doclint>-Xdoclint:none -notimestamp</doclint>
-          <additionalOptions>
-            <additionalOption>-Xdoclint:none -notimestamp</additionalOption>
-          </additionalOptions>
+          <doclint>none</doclint>
           <subpackages>org.apache.bookkeeper.client:org.apache.bookkeeper.client.api:org.apache.bookkeeper.common.annotation:org.apache.bookkeeper.conf:org.apache.bookkeeper.feature:org.apache.bookkeeper.stats</subpackages>
           <groups>
             <group>

--- a/stream/distributedlog/pom.xml
+++ b/stream/distributedlog/pom.xml
@@ -45,7 +45,8 @@
         <version>${maven-javadoc-plugin.version}</version>
         <configuration>
           <!-- Avoid for missing javadoc comments to be marked as errors -->
-          <doclint>-Xdoclint:none -notimestamp</doclint>
+          <additionalparam>-notimestamp</additionalparam>
+          <doclint>none</doclint>
           <groups>
             <group>
               <title>Core Library</title>

